### ISSUE TITLE
Could not find any entity of type <Entity> when using specific commands.

### ISSUE THREAD

**Issue by Jye-Dewar** [typeorm/typeorm#8040]
## Issue Description
Queries work when I use either QueryBuilder or Raw SQL, using find, findOne, getOne etc returns nothing.

### Expected Behavior
The any query created with the following would return the expected user:
```typescript
await db.getRepository<User>('User').findOneOrFail({ emailAddress });

// return
User {
firstName 'Foo',
lastName: 'Bar',
emailAddress: 'foo@bar.com'
}
```

### Actual Behavior
```typescript
await db.getRepository<User>('User').findOneOrFail({ emailAddress });
// "Could not find any entity of type \"User\" matching: {\n    \"emailAddress\": \"foo@bar.com\"\n}"

await db.getRepository<User>('User').createQueryBuilder()
    .where('emailAddress = :emailAddress', { emailAddress: `${emailAddress}` })
    .getOne();
// returns undefined

await db.getRepository<User>('User').createQueryBuilder()
    .where('emailAddress = :emailAddress', { emailAddress: `${emailAddress}` })
    .execute();
// works
```

### Steps to Reproduce
Code was working fine locally and deployed, before completing a new deployment. Upon completing a new deployment (none of the above code was changed), the error began appearing.
The error transferred to my local environment when I performed a fresh node_modules install (It was not happening until that happened, regardless of the status of the deployed version).

1.  Calling any find/findOne/getOne/findOneOrFail command reproduces the issue

```typescript
@Entity('user')
export class User {
  @PrimaryGeneratedColumn()
  userId: number;

  @Column({ type: 'varchar', unique: true, length: 255 })
  emailAddress: string;

  @Column({ type: 'varchar', default: '' })
  phoneNumber: string;

  @Column({ type: 'varchar', default: '' })
  title: string;

  @Column('text')
  preferredName: string;

  @Column('text')
  firstName: string;

  @Column('text')
  lastName: string;

  @Column({ type: 'varchar', default: '' })
  profilePhotoURL: string;

  @Column('text')
  accountType: string;
}

export const getUserQuery = async (emailAddress: string) => {
  const db = await initDb(`${process.env.NODE_ENV}-db`); // Even hard-coding the selected DB makes no impact, have checked that.
  return db.getRepository(User).findOneOrFail({ emailAddress });
```

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Operating System | Ubuntu 20.04  |
| Node.js version     | v12.16.3  |
| Typescript version  | 3.9.10 |
| TypeORM version     | 0.2.36  |

### Relevant Database Driver(s)

| DB Type              | Reproducible  |
| ---                  | ---           |
| `aurora-data-api`    | yes|
| `aurora-data-api-pg` | no            |
| `better-sqlite3`     | no            |
| `cockroachdb`        | no            |
| `cordova`            | no            |
| `expo`               | no            |
| `mongodb`            | no            |
| `mysql`              | no            |
| `nativescript`       | no            |
| `oracle`             | no            |
| `postgres`           | no            |
| `react-native`       | no            |
| `sap`                | no            |
| `sqlite`             | no            |
| `sqlite-abstract`    | no            |
| `sqljs`              | no            |
| `sqlserver`          | no            |

### Are you willing to resolve this issue by submitting a Pull Request?

 - ✖️ Yes, I have the time, and I know how to start.
 - ✅ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✅ No, I don’t have the time, but I can support (using donations) development.
 - ✖️ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

**Comment by blazmrak** [comment]
What is the query that gets executed?

**Comment by Jye-Dewar** [comment]
Pinning exact version resolves issue. 
```JSON
"typeorm": "^0.2.34" // Bug is present
"typeorm": "0.2.34" // Bug goes away
```

**Comment by jgomes94** [comment]
I can confirm this I also had this problem when using Aurora Data API driver.

**Comment by pleerock** [comment]
Can you please enable `logging` by setting `logging: true` in connection options and show us SQL query generated? Also, if you want issue to be resolved faster, good to have a git repository with smallest reproduction possible.

**Comment by Evandro18** [comment]
I'm using Postgres and I got the same problem.

**Comment by raphaelluchini** [comment]
I am facing this issue with Postgress as well. 
My tests fail when I try to update TypeOrm to ^0.2.45

@Jye-Dewar Yes, pinning works, but I want to update TypeOrm to the latest 2.x and later 3.x and that is blocking me from doing that :(

**Comment by blazmrak** [comment]
@raphaelluchini can you please provide the query that gets executed? (logging: true) Or minimal reproduction repo?

**Comment by raphaelluchini** [comment]
@blazmrak  Unfortunately the query is quite complex, I will try to simplify and maybe get a minimal example soon.

**Comment by blazmrak** [comment]
@raphaelluchini what happens if you run this query to DB directly?

**Comment by raphaelluchini** [comment]
@blazmrak 
Let me try to compare the queries between the versions. But apparently, the query is correct. The problem occurs when it does not return anything. Give me some time and can explain the issue better. ;)

**Comment by blazmrak** [comment]
@raphaelluchini Might be something with the mapping then... Minimal reproduction example would be very nice here. If you create a repo with this script it would be much appreciated:

test.ts
```javascript
// Entity
// Create connection
// Create table if not exists
// Insert into table
// Query one
```

**Comment by raphaelluchini** [comment]
(It's not fully related to this issue)
@blazmrak I am currently checking my problem and I am facing a different issue on top of that. Once I receive the `Could not find any entity of type User` the code is throwing on purpose EntityNotFoundError error.

I noticed we have an exception filter catching some types and that stopped working after I moved versions.
My service returns some specific gRPC error depending on the Error returned by TypeORM:

```ts
if (exception instanceof EntityNotFoundError) {
      const notFoundException = this.notFoundExceptionFactory();
      apiException = notFoundException.apiException;
      grpcException.code = notFoundException.grpcExceptionCode;
}
if (exception instanceof QueryFailedError) {
      const notFoundException = this.queryFailExceptionFactory(exception);
      apiException = notFoundException.apiException;
      grpcException.code = notFoundException.grpcExceptionCode;
}
```
The problem is that my ``exception`` is not matching `EntityNotFoundError`.
Here is the output of ``exception``:
```
EntityNotFoundError: Could not find any entity of type \"User\" matching: \"otp.token\"\n    at 
EntityNotFoundError.TypeORMError [as constructor] (.yarn/__virtual__/typeorm-virtual-529bd965bc/0/cache/typeorm-npm-0.2.45-8264084ef9-b7684a52c8.zip/src/error/TypeORMError.ts:7:9)\n    at
 new EntityNotFoundError (.yarn/__virtual__/typeorm-virtual-529bd965bc/0/cache/typeorm-npm-0.2.45-8264084ef9-b7684a52c8.zip/src/error/EntityNotFoundError.ts:10:9)\n...
```
I also have seen in the TypeORM source code that `EntityNotFoundError` now extends a new `TypeORMError`.

Do you have any clue here?

Thanks

**Comment by blazmrak** [comment]
@raphaelluchini Open a new issue for this, just so we keep this on topic.

**Comment by apatryda** [comment]
This happens because `createQueryBuilder` uses the class name when searching for the appropriate SQL table name while it should use an entity metadata entry instead.

**Comment by eilinora** [comment]
This is still an issue even with latest versions

**Comment by mehtsaur-amazon** [comment]
@apatryda so what is the workaround then? Should we start using table names as entity names?

**Comment by msau9** [comment]
@apatryda is there a workaround for this? Should we use table names as our class names? I tried that but it still shows me the same error.

**Comment by apatryda** [comment]
@msau9 - I just avoid passing the class to `createQueryBuilder` directly.

If I had an entity class `User`, instead of doing:

```typescript
const builder = entityManager.createQueryBuilder(User, 'u');
```

I'd do:

```typescript
const userMetadata = dataSource.getMetadata(User);
const builder = entityManager.createQueryBuilder(userMetadata.tableName, 'u');
```

Let me know if this helps.

**Comment by msau9** [comment]
Thanks for the quick response!

It still throws the same error with that method. I'm using typeorm with **NestJS** BTW.


```typescript
import { Repository } from 'typeorm';
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
...
...
@Injectable
export class MiscService {
  constructor( // Some boilerplate for dependency injection in NestJS, can ignore
    @InjectRepository(ActivityCategory) private categoryRepository: Repository<ActivityCategory>,
  ) { }
  ...
  ...
  getCategories() {
    const queryBuilder = this.categoryRepository.manager.createQueryBuilder(this.categoryRepository.metadata.tableName, 'category')
    return await queryBuilder.getOneOrFail(); 
  }
}
```

`EntityNotFoundError: Could not find any entity of type "ActivityCategory" matching: [object Object]`


I'm using the version `^0.3.11` and my tables are on MS SQL server.

**Comment by apatryda** [comment]
@msau9 - are there any records in that table?

**Comment by msau9** [comment]
Yep. I verified that the `getRawMany()` on the queryBuilder returns non-zero records.

**Comment by Nizarius** [comment]
Same problem with PostgreSQL driver

**Comment by vicary** [comment]
Experiencing this issue in v0.3, using Data API.

It happens when selecting right after deleting.

**Comment by VladimirFilipovic** [comment]
> Experiencing this issue in v0.3, using Data API.
> 
> It happens when selecting right after deleting.

Make sure that you are not selecting those row who are soft-deleted or deleted

**Comment by AmondChocolateShake** [comment]
You need to set the column names using "_", not Using camel case
